### PR TITLE
Adding expand_message_xmd and RFC compliance with VOPRF and H2C

### PR DESCRIPTION
**Pull request by kevinlewi**
Addresses #72, #73, and #74 

Note that x25519 is not in the VOPRF spec, and so we actually don't have anything to match against... but this will handle ristretto255 at least.

### CODE REVIEW

File: src/map_to_curve.rs [hunk 1]
```diff
@@ -6,31 +6,234 @@
...
+        Ok(<Self as Group>::hash_to_curve(&hashed_input))
+    }
+}
+
+// Computes ceil(x / y)
+fn div_ceil(x: usize, y: usize) -> usize {
+    let additive = match x % y != 0 {
+        true => 1,
+        false => 0,
+    };
```
Comment (huitseeker) [inline src/map_to_curve.rs:62]: Would `if x % y == 0 { 0 } else { 1 }` be simpler? (or if we're really going for it, `(x%y != 0) as usize`)

File: src/map_to_curve.rs [hunk 2]
```diff
@@ -6,31 +6,234 @@
...
+#[cfg(test)]
+mod tests {
+
+    struct Params {
+        msg: &'static str,
+        len_in_bytes: usize,
+        uniform_bytes: &'static str,
+    }
+
+    #[test]
```
Comment (huitseeker) [inline src/map_to_curve.rs:126]: Super!

File: src/serialization/mod.rs [hunk 3]
```diff
@@ -80,23 +80,43 @@ impl<T: CipherSuite> From<&LoginThirdMessage<T>> for ProtocolMessageType {
...
 }
 
-pub(crate) fn serialize(input: &[u8], max_bytes: usize) -> Vec<u8> {
-    let mut output: Vec<u8> = Vec::new();
-    output.extend_from_slice(&input.len().to_be_bytes()[8 - max_bytes..]);
-    output.extend_from_slice(&input[..]);
+// Corresponds to the I2OSP() function from RFC8017
+pub(crate) fn i2osp(input: usize, length: usize) -> Vec<u8> {
+    if length <= 8 {
+        return (&input.to_be_bytes()[8 - length..]).to_vec();
```
Comment (huitseeker) [inline src/serialization/mod.rs:86]: If usize is 32 bytes on a particular platform, and I want to write it on len=4
bytes, it seems I'd be writing .. nothing? `std::mem::size_of::<usize>()` may help here.

File: src/serialization/mod.rs [hunk 4]
```diff
@@ -80,23 +80,43 @@ impl<T: CipherSuite> From<&LoginThirdMessage<T>> for ProtocolMessageType {
...
 
+// Corresponds to the OS2IP() function from RFC8017
+pub(crate) fn os2ip(input: &[u8]) -> Result<usize, PakeError> {
+    if input.len() > 8 {
+        return Err(PakeError::SerializationError);
+    }
+
+    let mut output_array = [0u8; 8];
+    for i in 0..input.len() {
+        output_array[8 - input.len() + i] = input[i];
```
Comment (huitseeker) [inline src/serialization/mod.rs:102]: Would it be simpler to use `copy_from_slice`, considering it takes a `&mut self`?

File: src/serialization/mod.rs [hunk 5]
```diff
@@ -80,23 +80,43 @@ impl<T: CipherSuite> From<&LoginThirdMessage<T>> for ProtocolMessageType {
...
+    }
+    Ok(usize::from_be_bytes(output_array))
+}
```
Comment (huitseeker) [inline src/serialization/mod.rs:105]: Could we unit-test i2osp, is2ip, preferably with proptest and potentially with a cross-test on u32? I'm happy adding 32-bits compilation / testing to CI

**Comment by huitseeker** [comment]
@kevinlewi See also the run results on #81 which show we already have issues in this area
https://github.com/novifinancial/opaque-ke/pull/81/checks?check_run_id=1374926279

File: src/serialization/mod.rs (hunk 5 above)
Comment (huitseeker) [reply src/serialization/mod.rs:105]: CI is in #81

File: src/serialization/mod.rs (hunk 3 above)
Comment (kevinlewi) [reply src/serialization/mod.rs:86]: Changed "8" to std::mem::size_of::<usize>() everywhere

File: src/serialization/mod.rs (hunk 5 above)
Comment (kevinlewi) [reply src/serialization/mod.rs:105]: Added a proptest:

```
#[test]
fn test_i2osp_os2ip(bytes in vec(any::<u8>(), 0..std::mem::size_of::<usize>())) {
    assert_eq!(i2osp(os2ip(&bytes)?, bytes.len()), bytes);
}
```

**Review by huitseeker — APPROVED** [review]
Awesome!

**Pull request merged**